### ISSUE TITLE
A frontend to handle Velociraptor/TreeFrog mergertree files

### ISSUE THREAD

**Issue by manodeep** [ytree-project/ytree#66]
Here are the source repos for [Velociraptor](https://github.com/pelahi/VELOCIraptor-STF) and [TreeFrog](https://github.com/pelahi/TreeFrog)

The format comes with one file that contains forest-level information across all datafiles (`<prefix>.foreststats.hdf5`), and then the individual datafiles (`<prefix>.forestID.hdf5.[0-N]`) . 

The data format is reasonably standard - every file contains groups as "Snap_XXX" which in turn contain the relevant datasets for each halo property. All halos across all forests are stored contiguously, and there are two fields `/ForestInfoInFile/Snap_XXX/ForestOffsetPerSnap` and `/ForestInfoInFile/Snap_XXX/NHalosPerForestInSnap` of shape `(nforests_in_file, )` that dictate where to start loading and how many halos to load for any forest at a given snapshot. This datafile also contains the number of forests in that file  (`/ForestInfoInFile/NForestsInFile`) and the total number of halos in each forest across all snapshots (`/ForestInfoInFile/ForestSizesInFile`).

The `foreststats` file contains the total number of forests across all files (`/ForestInfo/NForests`), the number of files that the forests are spread across (`/Header/NFiles`) and the total number of snapshots (`/Header/NSnaps`). 

Here is the output from `h5dump -n 1 <filename>`: 
- [velociraptor.foreststats.hdf5.txt](https://github.com/ytree-project/ytree/files/5061000/velociraptor.foreststats.hdf5.txt) -- - contains aggregate info about forests across all files
- [velociraptor.forestID.hdf5.0.txt](https://github.com/ytree-project/ytree/files/5061001/velociraptor.forestID.hdf5.0.txt)  --  contains aggregate info about forests in that file, and all the halos (written out in separate snapshot groups) associated with all the forests in that file. Also contains the simulation info, cosmological params (`/Header/Simulation/Omega_Lambda` etc) and the units (`/Header/Simulation/Length_unit` etc)

To load a single forest, you would need to get to the correct offset (for all requested columns), and then load the relevant number of halos at each snapshot, and flatten out the array. The progenitor-descendant linking is done through IDs that can be translated into a snapshot-local array index with the following `C` code:

```C
#define CONVERSION_FACTOR_FOR_GENESIS_UNIQUE_INDEX      (1000000000000)
#define CONVERT_HALOID_TO_SNAPSHOT(haloid)    (haloid / CONVERSION_FACTOR_FOR_GENESIS_UNIQUE_INDEX )
#define CONVERT_HALOID_TO_INDEX(haloid)       ((haloid %  CONVERSION_FACTOR_FOR_GENESIS_UNIQUE_INDEX) - 1)
#define CONVERT_SNAPSHOT_AND_INDEX_TO_HALOID(snap, index)    (snap*CONVERSION_FACTOR_FOR_GENESIS_UNIQUE_INDEX + index + 1)
```

You will likely need more information but perhaps this is sufficient to gauge the complexity of the implementation. 

In case this is useful, [here](https://github.com/jacobseiler/sage-model/blob/genesis/src/io/read_tree_genesis_hdf5.c) is my (untested) implementation in `C`.

**Comment by manodeep** [comment]
Tagging @pelahi and @rhyspoulton (please feel free to unsubscribe).

**Comment by brittonsmith** [comment]
@manodeep, any chance you have some sample data in this format? Some about ~1 GB or less would be ideal, but anything would be useful. I might be ready to make a try at this.

**Comment by manodeep** [comment]
Thanks @brittonsmith. The smallest dataset I have is 1.7 GB - let me check if there is anything smaller floating around

**Comment by manodeep** [comment]
Here's a link to a [dropbox folder](https://www.dropbox.com/sh/535rnpilumegi3w/AACSzeL4R3140HM9gym8fYTza?dl=0) containing the two files - one with the actual data and the other with the metadata

**Comment by manodeep** [comment]
@brittonsmith Let me know if those files work for you. If the file size is an issue, please let me know - I can probably create a new file with a subset of the trees

**Comment by brittonsmith** [comment]
@manodeep, thanks, these should work fine for now. If it's possible to make something smaller, that would be better to use as test data. Currently, the whole test data collection is only about 1.6 GB. If it's a pain, though, don't worry about.

**Comment by manodeep** [comment]
I have to write a new script to only keep some fraction of the trees - might take me >~ a few weeks to get that sorted though. I am assuming the bigger files are fine for now while you develop the reader.

**Comment by brittonsmith** [comment]
Yeah, they'll be fine for that. It will probably be a similar amount of time for me to get to this.

**Comment by brittonsmith** [comment]
Resolved in PR #88.
